providertest.go

```diff
@@ -31,6 +31,7 @@ type ProviderTest struct {
 	config           map[string]string
 	e2eOptions       []E2eOption
 	skipSdk          map[string][]any
+	upgradeOpts      providerUpgradeOpts
```

Comment:
This is slightly controversial. My take on it is that the public API should be as simple as possible; instead of distinguishing various flavors of option XOption and Option at compile time we can get away with having one big Option type for now. The implementations can be fixed up without changing the interface to emit warnings if the user passes Option values that are non-actionable for a particular kind of test.

Open to other ways of doing this. 

For VerifyUpgrade, some options like `config` are needed but it also has some specific options of its own.

Reply:
An alternative here is dedicated VerifyUpgradeOption, but then we need to decide if WithConfig should be an Option or a VerifyUpgradeOption; stretching the type system further, VerifyUpgradeOption and Option could be interfaces and WithConfig could satisfy both.

Reply:
I think I agree with your philosophy here, but I don't follow why we need this struct to know about `upgradeOpts` at all in the current shape?

Reply:
It's a technicality.. because Option is defined as `func (*ProgramTest){}`.